### PR TITLE
Add grad_zero for phi and test function

### PR DESCRIPTION
**Pull request by arovinelli**
Just add two constants gradient variable

@lindsayad as per mailing list  discussion

ref #15204

### CODE REVIEW

File: framework/src/problems/FEProblemBase.C [hunk 1]
```diff
@@ -1600,6 +1604,8 @@ FEProblemBase::reinitDirac(const Elem * elem, THREAD_ID tid)
         _scalar_zero[tid].resize(FORTYTHIRD, 0);
         _zero[tid].resize(max_qpts, 0);
         _grad_zero[tid].resize(max_qpts, RealGradient(0.));
+        _grad_phi_zero[tid].resize(max_qpts, {0., 0., 0.});
+        _grad_test_zero[tid].resize(max_qpts, {0., 0., 0.});
```
Comment (arovinelli) [inline framework/src/problems/FEProblemBase.C:1608]: @lindsayad this init here is ugly but I can't figure out how to init with RealVectorValue...
suggestions?

**Comment by lindsayad** [comment]
I need to see a test to believe this is working. For one thing the outer size should be according to the number of shape functions, not the number of q-points. Perhaps you are getting away with this because for first-order Lagrange those numbers are the same in the element volume? Also my current reading of the `{0,0,0}` initializer list is that its going to create three `RealVectorValues` with each component initialized to zero. That length of three doesn't seem like it's going to match an arbitrary number of quadrature points. What are you testing this on?

**Comment by arovinelli** [comment]
> I need to see a test to believe this is working. For one thing the outer size should be according to the number of shape functions, not the number of q-points. Perhaps you are getting away with this because for first-order Lagrange those numbers are the same in the element volume? Also my current reading of the `{0,0,0}` initializer list is that its going to create three `RealVectorValues` with each component initialized to zero. That length of three doesn't seem like it's going to match an arbitrary number of quadrature points. What are you testing this on?

wait, that is not how the already available `_grad_zero` works. If I understand the avaialble code correctly, it is simply a `_grad_zero` is simply a RealVectorValue(0) with a size equal to the number of `_qp` such that you can do something like in this [example](https://github.com/idaholab/moose/blob/795302412e39a99768c0fe2c5a869c8cb3b17923/modules/porous_flow/src/materials/PorousFlowVolumetricStrain.C#L58-L75)
So you just need _grad_phi_zero to be a constant reference to a RealVector.Am I wrong?

I'm not testing it right now, try to make a kernel or material for this

**Comment by lindsayad** [comment]
A `VariableGradient` is a `MooseArray<RealVectorValue>`. A `VariablePhiGradient` is a `MooseArray<std::vector<RealVectorValue>`. So the latter takes two indexes, the first for dofs, the second for quadrature points. The former only takes a single index for quadrature points.

**Comment by arovinelli** [comment]
> A `VariableGradient` is a `MooseArray<RealVectorValue>`. A `VariablePhiGradient` is a `MooseArray<std::vector<RealVectorValue>`. So the latter takes two indexes, the first for dofs, the second for quadrature points. The former only takes a single index for quadrature points.

This make a lot of sense and explains why I was not able to find the right initialization for it.
So it should something like 
```
    _grad_test_zero[tid].resize(getMaxShapeFunctions(),
                                std::vector<RealGradient>(max_qpts, RealGradient(0.)));
```

quick question

**Comment by lindsayad** [comment]
Correct. You might actually want to look at the code around this line: https://github.com/idaholab/moose/blob/next/framework/src/problems/FEProblemBase.C#L627. We actually compute exactly what you're looking for there

**Comment by arovinelli** [comment]
@lindsay I'm also trying to construct a `VariapliPhiValue _phi_zero`. What type is and what size should a VariapliPhiValue ?

**Comment by lindsayad** [comment]
A `VariablePhiValue` is a `MooseArray<std::vector<Real>`.

**Comment by arovinelli** [comment]
@lindsay it isn't as straightforward as I thought. 
first, the number of DOF might be different between nodes and elements `_nl->getMaxVarNDofsPerElem`  `_nl->getMaxVarNDofsPerNode` so I have to choose which of this to use. Also when called in ` FEProbelmBase::createQRules` they both return zero, therefore  I assume ` FEProbelmBase::intialSetup` has not been called yet.

I tried to use the code you indicated , inside `FEProbelmBase::createQRules` but it gives me a segfault
```
    MaxVarNDofsPerElem mvndpe(*this, *_nl);
    Threads::parallel_reduce(*_mesh.getActiveLocalElementRange(), mvndpe);
    auto max_var_n_dofs_per_elem = mvndpe.max();
    _communicator.max(max_var_n_dofs_per_elem);
```

here is the stack trace
```
#0  0x00007fffe8596a22 in __gnu_debug::_Error_formatter::_Parameter::_M_print_field(__gnu_debug::_Error_formatter const*, char const*) const () from /lib64/libstdc++.so.6
#1  0x00007fffe8596cfb in __gnu_debug::_Error_formatter::_Parameter::_M_print_description(__gnu_debug::_Error_formatter const*) const () from /lib64/libstdc++.so.6
#2  0x00007fffe859738f in __gnu_debug::_Error_formatter::_M_error() const () from /lib64/libstdc++.so.6
#3  0x00007ffff743ca0b in std::__debug::vector<unsigned int, std::allocator<unsigned int> >::operator[] (this=0x892f10, __n=0)
    at /opt/rh/devtoolset-7/root/usr/include/c++/7/debug/vector:424
#4  0x00007fffeef8a9e5 in libMesh::DofMap::dof_indices (this=0x892e80, elem=0x85b650, 
    di=std::__debug::vector of length 0, capacity 0, vn=0, p_level=0)
    at /home/arovinelli/projects/NEW_stuff/moose/scripts/../libmesh/src/base/dof_map.C:2093
#5  0x00007ffff5599308 in MaxVarNDofsPerElem::onElement (this=0x7fffffffb930, elem=0x85b650)
    at /home/arovinelli/projects/NEW_stuff/moose/framework/src/base/MaxVarNDofsPerElem.C:39
#6  0x00007ffff47ba86b in ThreadedElementLoopBase<libMesh::StoredRange<libMesh::MeshBase::const_element_iterator, libMesh::Elem const*> >::operator() (this=0x7fffffffb930, range=..., bypass_threading=false)
    at /home/arovinelli/projects/NEW_stuff/moose/framework/build/header_symlinks/ThreadedElementLoopBase.h:209
#7  0x00007ffff4d4b073 in libMesh::Threads::parallel_reduce<libMesh::StoredRange<libMesh::MeshBase::const_element_iterator, libMesh::Elem const*>, MaxVarNDofsPerElem> (range=..., body=...)
    at /home/arovinelli/projects/NEW_stuff/moose/scripts/../libmesh/installed/include/libmesh/threads_pthread.h:380
#8  0x00007ffff4d21b23 in FEProblemBase::createQRules (this=0x8685a0, type=libMesh::QGAUSS, order=libMesh::THIRD, 
    volume_order=libMesh::THIRD, face_order=libMesh::THIRD)
    at /home/arovinelli/projects/NEW_stuff/moose/framework/src/problems/FEProblemBase.C:4580
#9  0x00007ffff4c676a2 in SetupQuadratureAction::act (this=0x81fc90)
    at /home/arovinelli/projects/NEW_stuff/moose/framework/src/actions/SetupQuadratureAction.C:51
```
so I can't compute either the maxNdof per elem or node

**Comment by lindsayad** [comment]
It might be helpful for you to look at the order of actions in `Moose.C`, starting around line 245. Quadrature setup happens well before the `FEProblemBase` is initialized (the `init_problem` task calls `FEProblemBase::init`), which is where the libMesh `EquationSystems` and `DofMap` objects are initialized, which is where dof indices and things will get setup. 

Why don't you wait do your zeroing and setup during `FEProblemBase::initialSetup` which is where we already do those computations? You can choose to make that calculation non-optional if you want.

**Comment by arovinelli** [comment]
@lindsay, thanks a lot for the suggestion I've embraced it and I'm making good progress.
At this point, I believe the size of _second_phi_zero wrong and also initialized in the wrong place, could you check (see link) 
https://github.com/idaholab/moose/blob/544d67eb7d02227ed071bd74b677aeb5674c8d24/framework/src/problems/FEProblemBase.C#L1612-L1613

**Comment by lindsayad** [comment]
Looking at MaxQpsThread::operator(), `getMaxShapeFunctions` will definitely
return the wrong value if there are second order bases.

**Comment by arovinelli** [comment]
So this is a preexisting bug.
This PR should fix it.

> Looking at MaxQpsThread::operator(), `getMaxShapeFunctions` will definitely return the wrong value if there are second order bases.
> […](#)

File: test/src/kernels/PhiZeroKernel.C [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+InputParameters
+PhiZeroKernel::validParams()
+{
+  InputParameters params = NullKernel::validParams();
+  params.addClassDescription("Test for phi zero values");
+  return params;
+}
+
+PhiZeroKernel::PhiZeroKernel(const InputParameters & parameters)
+  : NullKernel(parameters), _second_phi(_assembly.secondPhi(_var))
```
Comment (arovinelli) [inline test/src/kernels/PhiZeroKernel.C:23]: Why I can't get _second_phi.size() > 0? In my tests, I use second-order element and second-order lagrangian function

Comment (lindsayad) [reply test/src/kernels/PhiZeroKernel.C:23]: Just calling `secondPhi` in `Assembly` is not enough to get MOOSE to do second derivatives. You need to actually request something like `_var.secondSln()` in order to get it to happen. A hacky way that may also work is to directly call `_assembly.feSecondPhi<Real>(FEType(LAGRANGE,FIRST))`

Comment (arovinelli) [reply test/src/kernels/PhiZeroKernel.C]: thanks, now it works!

**Comment by arovinelli** [comment]
@lindsayad when you have time, this is ready for review

File: framework/src/problems/FEProblemBase.C [hunk 3]
```diff
@@ -626,40 +630,46 @@ FEProblemBase::initialSetup()
   setCurrentExecuteOnFlag(EXEC_INITIAL);
   addExtraVectors();
 
-  // Execute this here in case we want to print out the required derivative size in
-  // OutputWarehouse::initialSetup
-  if (haveADObjects() || (_displaced_problem && _displaced_problem->haveADObjects()))
-  {
-    CONSOLE_TIMED_PRINT("Computing max dofs per elem/node");
+  // alwasy esecute to get the max number of DoF per elment and node needed to initialized phi_zero
```
Comment (lindsayad) [inline framework/src/problems/FEProblemBase.C:633]: always execute

File: test/include/kernels/PhiZeroKernel.h [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#include "NullKernel.h"
+
+class PhiZeroKernel;
```
Comment (lindsayad) [inline test/include/kernels/PhiZeroKernel.h:14]: No need for this forward declaration

File: test/include/kernels/PhiZeroKernel.h [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+
+/**
+ *
+ */
```
Comment (lindsayad) [inline test/include/kernels/PhiZeroKernel.h:18]: Either add a doxygen comment or just delete this

File: test/src/kernels/PhiZeroKernel.C [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+
+Real
+PhiZeroKernel::computeQpResidual()
+{
+
+  if (_phi.size() > 0)
+  {
+    mooseAssert(_phi.size() == _phi_zero.size(),
+                "_phi.size() " + std::to_string(_phi.size()) + "> _phi_zero.size() " +
+                    std::to_string(_phi_zero.size()));
```
Comment (lindsayad) [inline test/src/kernels/PhiZeroKernel.C:36]: I understand that you were thinking about cases in which the "zero" members were of zero size, but your message should probably just say "!=" as opposed to ">". Same for all the following assertions

File: test/tests/phi_zero/tests [hunk 7]
```diff
@@ -0,0 +1,22 @@
+[Tests]
+  [./phi_zero_linear]
+    type = 'RunApp'
+    input = 'simple_transient_diffusion.i'
+    expect_out = ' Solve Converged!'
+    issues = '#15204'
+    design = ' '
```
Comment (lindsayad) [inline test/tests/phi_zero/tests:7]: ruh roh. You should either write some documentation or find some existing documentation that roughly fits what you're working on

File: test/tests/phi_zero/tests [hunk 8]
```diff
@@ -0,0 +1,22 @@
+[Tests]
+  [./phi_zero_linear]
+    type = 'RunApp'
+    input = 'simple_transient_diffusion.i'
+    expect_out = ' Solve Converged!'
```
Comment (lindsayad) [inline test/tests/phi_zero/tests:5]: You can remove this. If the solve didn't converge, then this test would register as failed

File: test/tests/phi_zero/tests [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+    issues = '#15204'
+    design = ' '
+    requirement = 'The system shall be able to produce consestent size _phi_zero and _grad_phi_zero objects'
```
Comment (lindsayad) [inline test/tests/phi_zero/tests:8]: We want our requirements to be as agnostic of implementation as possible. Maybe something more like:
```
requirement = 'The system shall be able to construct zero objects for shape functions and shape function gradients consistent with the maximum number of shape functions and quadrature points in the simulation.'
```

File: test/tests/phi_zero/tests [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+    method = 'dbg'
```
Comment (lindsayad) [inline test/tests/phi_zero/tests:9]: You might as well remove this as well. I understand that your assertions won't be tested, but the problem should still run for all methods

File: test/tests/phi_zero/tests [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+  [../]
+  [./phi_zero_qudratic]
+    type = 'RunApp'
+    input = 'simple_transient_diffusion.i'
+    cli_args = 'Mesh/elem_type=QUAD8  Variables/dummy/order=SECOND Variables/u/order=SECOND Kernels/phi_zero/second_order=true'
+    requirement = 'The system shall be able to produce consestent size _phi_zero and _grad_phi_zero'
+    'objects for quadratic shape functions'
+    expect_out = ' Solve Converged!'
```
Comment (lindsayad) [inline test/tests/phi_zero/tests:17]: Again can remove

File: test/tests/phi_zero/tests [hunk 12]
```diff
@@ -0,0 +1,22 @@
...
+    issues = '#15204'
+    design = ' '
```
Comment (lindsayad) [inline test/tests/phi_zero/tests:19]: need doc

File: test/tests/phi_zero/tests [hunk 13]
```diff
@@ -0,0 +1,22 @@
...
+    requirement = 'The system shall be able to produce consestent size _phi_zero and _grad_phi_zero objects'
+    method = 'dbg'
+  [../]
+  [./phi_zero_qudratic]
+    type = 'RunApp'
+    input = 'simple_transient_diffusion.i'
+    cli_args = 'Mesh/elem_type=QUAD8  Variables/dummy/order=SECOND Variables/u/order=SECOND Kernels/phi_zero/second_order=true'
+    requirement = 'The system shall be able to produce consestent size _phi_zero and _grad_phi_zero'
```
Comment (lindsayad) [inline test/tests/phi_zero/tests:15]: could use a similar requirement to the one I proposed above

File: test/tests/phi_zero/tests [hunk 14]
```diff
@@ -0,0 +1,22 @@
...
+    'objects for quadratic shape functions'
+    expect_out = ' Solve Converged!'
+    issues = '#15204'
+    design = ' '
+    method = 'dbg'
```
Comment (lindsayad) [inline test/tests/phi_zero/tests:20]: remove

**Review by lindsayad — CHANGES_REQUESTED** [review]
A nice addition. thank you @arovinelli . I'm curious what your use case is?

Mostly fixup the test specs file

**Comment by arovinelli** [comment]
> A nice addition. thank you @arovinelli . I'm curious what your use case is?
>

@lindsay This is a preparation step for implementing large deformation cohesive zone modeling. The main reason I needed this is to ease the formulation of the jacobian of the traction w.r.t. the discreet displacements. It might not be optimal in terms of efficiency but make the code much more understandable. It will also allow using exactly the same formulation for 1D 2D and 3D problems, thus avoiding a bunch of if statements that are detrimental when performing maintenance.
 
> Mostly fixup the test specs file

The last PR should address all your comments. Check the design and requirement parts to see if they comply with your specs.

Alex, thanks again for helping with this!!

**Comment by arovinelli** [comment]
I can't put the `.md` file in the test doc directory, as it will not be found. Suggestions?

**Comment by arovinelli** [comment]
@lindsay this has passed all tests, please give it another review and let me know if there is something left to do.

File: framework/doc/content/source/variables/PhiZero.md [hunk 15]
```diff
@@ -0,0 +1,6 @@
+# PhiZero
+
+MOOSE has helper zero objects for shape functions and shape function gradients that can be accessed by all MOOSE objects. The names of the zero objects are: `_phi_zero`, `_grad_phi_zero` and `_second_phi_zero`.
+These object first two dimensions are,: (i) the maximum number of degree of freedom per variable per element, and (ii) the maximum number of quadrature points per element.
```
Comment (lindsayad) [inline framework/doc/content/source/variables/PhiZero.md:4]: remove comma in `are,`

File: framework/doc/content/source/variables/PhiZero.md [hunk 16]
```diff
@@ -0,0 +1,6 @@
...
+The `PhiZeroKernel` checks that the first two dimensions of `_phi_zero`, `_grad_phi_zero` and `_second_phi_zero` are
```
Comment (lindsayad) [inline framework/doc/content/source/variables/PhiZero.md:5]: Probably say the "`PhiZeroKernel` test object"

**Review by lindsayad — CHANGES_REQUESTED** [review]
I really like how this looks. Just the two tiny doc requests, and then I'll approve!

**Review by lindsayad — APPROVED** [review]

**Pull request merged**